eth/db/storage.py

```diff
@@ -0,0 +1,477 @@
+                )
+
+    @property
+    def has_changed_root(self) -> bool:
+        return self._storage_lookup.has_changed_root
+
+    def get_changed_root(self) -> Hash32:
+        return self._storage_lookup.get_changed_root()
+
+    def persist(self, db: BaseDB) -> None:
+        self._validate_flushed()
+        if self._storage_lookup.has_changed_root:
+            self._storage_lookup.commit_to(db)
+            self._storage_lookup.clear_changed_root()
+
+
+class StorageDB(BaseStorageDB):
+    """
+    Caches lookups and batches writes to storage. One key goal is to only generate
+    storage roots at the end of the block (post-Byzantium).
+
+    Changes are persisted to a trie when :meth:`persist()` is called.
+    """
+    logger = cast(ExtendedDebugLogger, logging.getLogger("eth.db.storage.StorageDB"))
+
+    def __init__(self, db: BaseAtomicDB, account_db: BaseAccountDB) -> None:
+        """
+        Internal implementation details (subject to rapid change).
+
+        Each account is tracked through its own pipeline of databases,
+        which are responsible for journaling, batching, converting to a trie,
+        and so on.
+
+        We generate the pipelines on the fly, as they are needed for reading or writing.
+        We explicitly keep track of which accounts were written to, so that we can identify
+        which accounts might have changed storage roots.
+        """
+        self._account_db = account_db
+        self._raw_store_db = db
+        self._account_stores = {}  # type: Dict[Address, AccountStorageDB]
+        self._dirty_accounts = set()  # type: Set[Address]
+
+    def _get_address_store(self, address: Address) -> AccountStorageDB:
+        if address in self._account_stores:
+            store = self._account_stores[address]
+        else:
+            store = AccountStorageDB(self._raw_store_db, self._account_db, address)
+            self._account_stores[address] = store
+        return store
+
+    def _trigger_writeable_storage(self, address: Address) -> None:
+        self._dirty_accounts.add(address)
+
+    def _written_account_stores(self) -> Iterable[Tuple[Address, AccountStorageDB]]:
+        for address in self._dirty_accounts:
+            store = self._account_stores[address]
+            yield address, store
+
+    def get(self, address: Address, slot: int, from_journal: bool=True) -> int:
+        account_store = self._get_address_store(address)
+        return account_store.get(slot, from_journal)
+
+    def set(self, address: Address, slot: int, value: int) -> None:
+        account_store = self._get_address_store(address)
+        if address not in self._dirty_accounts:
+            self._trigger_writeable_storage(address)
+        account_store.set(slot, value)
+
+    def delete(self, address: Address) -> None:
+        account_store = self._get_address_store(address)
+        if address not in self._dirty_accounts:
+            self._trigger_writeable_storage(address)
```

Comment:
It was actually a little bit faster based on a quick `%timeit` run, and I was sort of ambivalent on the syntax. (Other than inlining, which I'm already an board with) If you think it's easier to read as one line, then I'll go ahead and do that. I'll assume you think so unless you comment otherwise.